aten/src/THC/THCAtomics.cuh

```diff
@@ -40,7 +42,10 @@ struct AtomicAddIntegerImpl<T, 2> {
 
     do {
       assumed = old;
-      sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);
+      sum = (size_t)address & 2 ? old >> 16 : old & 0xffff;
```

Comment:
Why are we casting `address` to `size_t`, but we're using `uint32_t` for it above? Does CUDA address memory using 32 bits?

Reply:
I was trying to change this as little as possible since it's a bit confusing, but, my understanding of this code is that we can only atomically compare-and-swap 32 bits of memory at a time, and only from a "round" memory addresses. What we're checking with `(size_t)address & 2` is if the *address* is odd/even (if it points to directly addressable memory or if we have to shift to CAS the value we want). I think `size_t` is just always going to be the size of the address, regardless of if it's 32 or 64 bit. The size of the address is unrelated to the size of the value stored at that address, and the value we actually care about is half (or a quarter) of the size of the smallest number of bits we can actually directly CAS.

The uint32_t is just an arbitrary 32 bit type to hold the smallest number of bits we can atomically compare-and-swap from the address. We want to avoid doing math on that uint32, since it's really holding one or more values in the format of `T`.

Reply:
Doing `(size_t)address & 2` is functionally equivalent to `address & 2` but potentially slower due to doing the computation in int64_t, as far as I can tell. Does it look different to you? If not we should take this chance to clean up the code anyways.

Reply:
I think you're commenting on an older version of this? I made some changes Friday